### PR TITLE
Make keyed object fragments an opaque type

### PR DESCRIPTION
**Pull request by sebmarkbage**
This triggers a warning if you try to pass a keyed object as a child.

You now have to wrap it in React.addons.createFragment(object) which
creates a proxy (in dev) which warns if it is accessed. The purpose of
this is to make these into opaque objects so that nobody relies on its
data structure.

After that we can turn it into a different data structure such as a
ReactFragment node or an iterable of flattened ReactElements.

### CODE REVIEW

**Comment by syranide** [comment]
Cool stuff! But no fragments from arrays?

**Comment by sebmarkbage** [comment]
This is really just an upgrade path. Arrays works just like they do today. However, you can use this to key two different arrays. React.addons.fragment({ a: [c1, c2], b: [c3, c4] }) so that you can swap the order of two whole arrays.

I'm thinking that we'd move to a model where we can key a fragment as an element. Something like:

```javascript
[<frag key="a">{[c1, c2]}</frag>, <frag key="b">{[c3, c4]}</frag>]
```

File: src/addons/ReactFragment.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    canWarnForReactFragment = true;
+  } catch (x) { }
+
+  var proxyPropertyAccessWithWarning = function(obj, key) {
+    Object.defineProperty(obj, key, {
+      enumerable: true,
+      get: function() {
+        warning(
+          false,
+          'A ReactFragment is an opaque type. You cannot access any of its ' +
```
Comment (spicyj) [inline src/addons/ReactFragment.js]: cannot? should not?

**Comment by syranide** [comment]
I'm asking because the array syntax suuuuuuucks to work with in JSX :)

File: src/addons/ReactFragment.js (hunk 1 above)
Comment (sebmarkbage) [inline src/addons/ReactFragment.js]: hm... Well technically you can because it is an upgrade path but we'll immediately break this next release so then you can't. Not sure how to phrase it to make it clear: Don't do that.

File: src/addons/ReactFragment.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      if (canWarnForReactFragment) {
+        var proxy = {};
+        Object.defineProperty(proxy, fragmentKey, {
+          enumerable: false,
+          value: object
+        });
+        for (var key in object) {
+          proxyPropertyAccessWithWarning(proxy, key);
+        }
+        Object.freeze(proxy);
```
Comment (spicyj) [inline src/addons/ReactFragment.js]: Just checking – adding new keys after creating a fragment will silently fail in non-strict mode?

**Comment by sebmarkbage** [comment]
@syranide What do you mean? Example of what you'd like to be able to do?

File: src/addons/ReactFragment.js (hunk 2 above)
Comment (sebmarkbage) [inline src/addons/ReactFragment.js]: Yea, which will be different in prod. It's the best we can do without real proxies. Let's just hope people use strict mode.

File: src/classic/element/ReactElementValidator.js [hunk 3]
```diff
@@ -213,19 +199,21 @@ function validateChildKeys(node, parentType) {
   } else if (node) {
     var iteratorFn = getIteratorFn(node);
     // Entry iterators provide implicit keys.
-    if (iteratorFn && iteratorFn !== node.entries) {
-      var iterator = iteratorFn.call(node);
-      var step;
-      while (!(step = iterator.next()).done) {
-        if (ReactElement.isValidElement(step.value)) {
-          validateExplicitKey(step.value, parentType);
+    if (iteratorFn) {
```
Comment (spicyj) [inline src/classic/element/ReactElementValidator.js]: Is this change related?

**Comment by syranide** [comment]
@sebmarkbage `<frag><div /><div /></frag>` (or w/e) ... `{[<div />, <div />]}` is not very nice.

File: src/core/__tests__/ReactMultiChildReconcile-test.js [hunk 4]
```diff
@@ -107,13 +107,14 @@ var FriendsStatusDisplay = React.createClass({
...
-    var children = null;
+    var children = [];
     var key;
     for (key in this.props.usernameToStatus) {
       var status = this.props.usernameToStatus[key];
-      children = children || {};
-      children[key] = !status ? null :
-          <StatusDisplay ref={key} status={status} />;
+      children.push(
+        !status ? null :
```
Comment (spicyj) [inline src/core/__tests__/ReactMultiChildReconcile-test.js]: Not sure what this test is testing, but perhaps it makes sense to skip the null completely?

File: src/core/__tests__/ReactMultiChildText-test.js [hunk 5]
```diff
@@ -16,10 +16,13 @@
 require('mock-modules');
 
 var React = require('React');
+var ReactFragment = require('ReactFragment');
 var ReactTestUtils = require('ReactTestUtils');
 
 var reactComponentExpect = require('reactComponentExpect');
 
+var frag = ReactFragment.create;
```
Comment (spicyj) [inline src/core/__tests__/ReactMultiChildText-test.js]: nit: any reason for inconsistency between this assignment and (in other tests) this:

```
function frag(obj) {
  return ReactFragment.create(obj);
}
```

?

File: src/utils/__tests__/ReactChildren-test.js [hunk 6]
```diff
@@ -94,7 +112,8 @@ describe('ReactChildren', function() {
       return this;
     };
 
-    var scopeTester = {};
+    // TODO: Use an object to test, after non-object fragments has fully landed.
```
Comment (spicyj) [inline src/utils/__tests__/ReactChildren-test.js]: Sorry, can you explain this change?

File: src/classic/element/ReactElementValidator.js (hunk 3 above)
Comment (sebmarkbage) [inline src/classic/element/ReactElementValidator.js]: Kind of. Without this change, a Map object would be iterated over as if it was a keyed object.

I fixed it so that plain objects are silently ignored by detecting if they have an element child. A heuristic. So I think that this should work again. However, we might as well ignore Maps since they're likely not fragments.

File: src/core/__tests__/ReactMultiChildText-test.js (hunk 5 above)
Comment (sebmarkbage) [inline src/core/__tests__/ReactMultiChildText-test.js]: Other tests require in beforeEach, so we're already inconsistent in that regard.

File: src/utils/__tests__/ReactChildren-test.js (hunk 6 above)
Comment (sebmarkbage) [inline src/utils/__tests__/ReactChildren-test.js]: I use the ReactChildren helpers to iterate over the result of a helper. Since an object is not terminal, yet, I can't use that to test. Once we fix this fragment thing, we can revert to just use the object here.

File: src/addons/ReactFragment.js (hunk 1 above)
Comment (spicyj) [inline src/addons/ReactFragment.js]: "Accessing its properties directly is deprecated and will not work in the next release."

perhaps. fine as-is too.

**Comment by spicyj** [comment]
lgtm

File: src/core/__tests__/ReactMultiChildReconcile-test.js (hunk 4 above)
Comment (sebmarkbage) [inline src/core/__tests__/ReactMultiChildReconcile-test.js]: Didn't want to change what the test does really.

Comment (spicyj) [inline src/core/__tests__/ReactMultiChildReconcile-test.js]: Well the null is implicitly keyed now so it's already changing some but sure.

**Comment by sebmarkbage** [comment]
@syranide I guess you can use this frag concept to solve that syntactic issue too. The primary purpose was to be able to provide a key to an unknown set. E.g. `<frag key="a">{this.props.children}</frag>` but I guess it would be neat to have both. E4X had syntax like this: `<><div /><div /></>` Not sure where to put the key there.

**Comment by sebmarkbage** [comment]
I'm going to need some kind of dedup logic for this. Otherwise it might be too spammy.

File: src/classic/element/ReactElementValidator.js (hunk 3 above)
Comment (leebyron) [inline src/classic/element/ReactElementValidator.js]: Note that this may break some use cases people have when using Map / OrderedMap with v13-beta/master. Nothing wrong with that, but something people need to be aware of.

Since Maps are iterable, and they yield `[key, value]` tuples, you'll start to see your Map keys showing up in the UI.

No clear fix with ES6 Map, unfortunately (not that anyone is using them in this way), but for Immutable.js Map you'll need to call `.toIndexedSeq().map()` instead of just `.map()` with the current behavior.

File: src/classic/element/__tests__/ReactElementValidator-test.js [hunk 7]
```diff
@@ -135,7 +141,7 @@ describe('ReactElementValidator', function() {
     spyOn(console, 'warn');
     var Component = React.createFactory(ComponentClass);
 
-    Component(null, { 1: Component(), 2: Component() });
+    Component(null, frag({ 1: Component(), 2: Component() }));
```
Comment (leebyron) [inline src/classic/element/__tests__/ReactElementValidator-test.js]: I foresee more of this pattern in the future since `ReactFragment.create()` is a little verbose. Maybe package this as `React.addons.frag`?

Comment (sebmarkbage) [inline src/classic/element/__tests__/ReactElementValidator-test.js]: It is a placeholder as an upgrade path, not a permanent API, which is why it is in addons. We're thinking something like:

`<frag><Component /><Component /></frag>`

or E4X's

`<><Component /><Component /></>`

Or when you want to key an unknown set:

`<div><frag key="a">{childrenA}</frag><frag key="b">{childrenB}</frag></div>`

`createFragment` is just a way to accept legacy object literals as an easy upgrade path.

File: src/classic/element/ReactElementValidator.js (hunk 3 above)
Comment (sebmarkbage) [inline src/classic/element/ReactElementValidator.js]: No, this is fixing that behavior. Note the next line that ignores validation of Maps. Previously it would try to validate the Map as if it was an object, which depending on the type of Map might fail.

(I'm not sure Maps (iterables of `[key, value]`) should be a supported way to key children. I think it makes more sense to turn Map data into iterables of ReactElement which is effectively a `[key, value]` tuple where key is `type + key`. But that is unrelated to this PR.)

**Comment by zpao** [comment]
No idea why it didn't show up, but Travis approves: https://travis-ci.org/facebook/react/builds/50125739

**Comment by zpao** [comment]
Ok, well, Travis approved before I made changes to lint rules. Now... not so much https://travis-ci.org/facebook/react/jobs/50128396

**Comment by sebmarkbage** [comment]
Thanks for trying tho, @coveralls.
